### PR TITLE
Added conditional GAN using PyTorch framework

### PR DESCRIPTION
**Pull request by ayush12gupta**

### CODE REVIEW

File: generative_models/cGAN_PyTorch/main.py [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+torch.backends.cudnn.benchmark = True
+import matplotlib.pyplot as plt
+from dataset import get_loader
+import imageio
+from model import *
+
+
+
+
+
```
Comment (nirmal-suthar) [inline generative_models/cGAN_PyTorch/main.py]: remove stray lines

File: generative_models/cGAN_PyTorch/main.py [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+def sample_image(args,z,labels, batches_done, generator):
+    """Saves a grid of generated digits ranging from 0 to n_classes"""
+    gen_imgs = generator(z, labels)
+    save_image(gen_imgs.data, args.outdir +"Image/%d.png" % batches_done, nrow=args.num_class, normalize=True)
+
```
Comment (nirmal-suthar) [inline generative_models/cGAN_PyTorch/main.py]: stray line

File: generative_models/cGAN_PyTorch/main.py [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+        sample_image(args,fix_noise,fix_label, epoch, generator)
+
+        # Checkpoint
+        torch.save(generator.state_dict(), args.outdir + 'generator_state/generator_{}_.pth'.format(epoch))
+        torch.save(discriminator.state_dict(),args.outdir + 'discriminator_state/discriminator_{}_.pth'.format(epoch))
+
+    return G_losses, D_losses
+
+def main():
```
Comment (nirmal-suthar) [inline generative_models/cGAN_PyTorch/main.py:25]: maybe position main() function above train() function, it is nice to have parser at the top for reference.

File: generative_models/cGAN_PyTorch/model.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
```
Comment (nirmal-suthar) [inline generative_models/cGAN_PyTorch/model.py:4]: stray lines

File: generative_models/cGAN_PyTorch/model.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+      def forward(self, img, label):
+          x = img.view(img.size(0), -1)
+          # print(x.shape)
+          z = self.label_embedding(label)
+          x = torch.cat([x, z], 1)
+          out = self.model(x)
+          return out
+
+
+
```
Comment (nirmal-suthar) [inline generative_models/cGAN_PyTorch/model.py]: stray lines

File: generative_models/cGAN_PyTorch/main.py [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+    parser = argparse.ArgumentParser()
+    # model config
+    parser.add_argument('--channel', type=int, default=3)
```
Comment (nirmal-suthar) [inline generative_models/cGAN_PyTorch/main.py]: channel argument is debatable as it also depends on dataset, so maybe add channel field according to dataset or remove it completely

File: generative_models/cGAN_PyTorch/README.md [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+```bash
+$ python3 main.py --ndata 'cifar10' --channel 3 --epochs 100
+```
+NOTE: on Colab Notebook use following command:
+```python
+!git clone link-to-repo
+%run main.py --ndata 'cifar10' --channel 3 --epochs 100 
+```
+### Reference
+https://arxiv.org/abs/1411.1784
```
Comment (nirmal-suthar) [inline generative_models/cGAN_PyTorch/README.md]: add the name of paper too, see references in other readme

**Review by nirmal-suthar — COMMENTED** [review]
Remove all the pre-trained models as it takes unnecessary space and make a folder named assets, and move all the display files (like gif and png) and adjust readme accordingly. Also, it would nice if you could compress gifs or simply use last image of gif for displaying, as gifs are taking lot of space (around 10 mb) Preferably use final epoch image for displaying, or use approx 5-8 images in gif.

File: generative_models/cGAN_PyTorch/main.py (hunk 6 above)
Comment (ayush12gupta) [reply generative_models/cGAN_PyTorch/main.py]: Made all the changes suggested

**Comment by ayush12gupta** [comment]
Made all the changes suggested

File: generative_models/cGAN_PyTorch/main.py [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+            rand_label = Variable(label_type(np.random.randint(0, args.num_class, batch_size))).to(args.device)
+
+            d_real = discriminator(imgs, labels)
+            loss_real = loss(d_real, r_label)
+
+            d_fake = discriminator(generator(noise, rand_label).detach(), rand_label)
+            # print(d_fake.shape," ",f_label.shape)
+            loss_fake = loss(d_fake, f_label)
+
+            d_loss = 0.5 * (loss_fake + loss_real)
```
Comment (nirmal-suthar) [inline generative_models/cGAN_PyTorch/main.py]:
```suggestion
            d_loss = loss_fake + loss_real
```

**Review by nirmal-suthar — APPROVED** [review]

File: generative_models/cGAN_PyTorch/main.py (hunk 8 above)
Comment (ayush12gupta) [reply generative_models/cGAN_PyTorch/main.py]: Do I need to train all these models again and generate the graph and all

Comment (ayush12gupta) [reply generative_models/cGAN_PyTorch/main.py]: uploaded new graphs for this commit

**Pull request merged**